doc/specs/#5000 - Process Model 2.0/#5000 - Process Model 2.0.md

```diff
@@ -0,0 +1,882 @@
+   process's `SwapChainPanel`.
+4. The content process will read output from the connection and draw to the swap chain. The
+   contents that are rendered to the swap chain will be visible in the window
+   process's `SwapChainPanel`, because they share the same underlying kernel
+   object.
+
+![figure-003](figure-003.png)
+
+The content process will be responsible for the terminal buffer and other
+terminal state (the core `Terminal` object), as well as the `Renderer`,
+`DxEngine`, and `TerminalConnection`. These are all being combined in the
+content process, as to maximize performance. We don't want to have to hop across
+the process boundary multiple times per frame, so the renderer must be in the
+same process as the buffer. Similarly, we want to be able to read data off of
+the connection as quickly as possible, and the best way to do this will be to
+have the connection in the same process as the `Terminal` core.
+
+
+
+#### Technical Details
+
+Much of the above is powered by the magic of WinRT (which is powered by the
+magic of COM). Whenever we create WinRT types, WinRT provides metadata about
+these types that not only enables us to use them in-proc (via the
+implementation), but also enables using these types _across process boundaries_.
+
+Typically, these classes are given a unique GUID for the class. A process that
+wants to implement one of these out-of-proc WinRT objects can register with the
+system to say "I make `MyClass`'s, and their GUID is `{foo}`". these are called
+WinRT _servers_. Then, consumers (_clients_) of that class can ask the system
+"I'd like to instantiate a `{foo}` please", which will cause the server to
+create a new instance of that object (in the server process), and the client
+will be given a winrt object which can interact with the classes WinRT
+projection.
+
+A server can register the types it produces with `CoRegisterClassObject`, like so:
+
+```c++
+CoRegisterClassObject(MyClassGUID,
+                      winrt::make<MyClassFactory>().get(),
+                      CLSCTX_LOCAL_SERVER,
+                      REGCLS_MULTIPLEUSE,
+                      &dwRegistration);
+
+```
+
+And a client can attempt to get an instance of `MyClass` with
+
+```c++
+auto myClass = create_instance<winrt::MyClass>(MyClassGUID, CLSCTX_LOCAL_SERVER);
+```
+
+We're going to be using that system a little differently here. Instead of using
+a GUID to represent a single _Class_, we're going to use the GUID to uniquely
+identify _content processes_. Each content process will receive a unique GUID
+when it is created, and it will register as the server for that GUID. Then, any
+window process will be able to connect to that specific content process strictly
+by GUID. Because each GUID is unique to each content process, any time any
+client calls `create_instance<>(theGuid, ...)`, it will uniquely attempt to
+connect to the content process hosting `theGuid`.

```

Comment:
I don't know how to feel about this. I can't name another example of something generating COM class IDs on the fly to represent content processes. Is there a source in browser land or some other multi-process content/frame management code that does this already?

If we're the first ones to do it, it makes me think that we're "holding it wrong" and there's an existing provision somewhere that is better for this than camping the COM catalog.

Reply:
Using COM for this is _probably_ unprecedented, which is why I'm gonna summon some more senior architects to take a look at this and tell me if I'm being batty. 

Just from examining commandlines associated with edgium and firefox, the idea of passing a unique ID on the commandline doesn't seem totally farfetched:

![image](https://user-images.githubusercontent.com/18356694/90032930-84cc5980-dc84-11ea-8720-cdc1990b33d1.png)
![image](https://user-images.githubusercontent.com/18356694/90032939-872eb380-dc84-11ea-889b-de2f00e67b68.png)

I'd presume they're just using a pipe or something for the communication, and I wanted something that might be more robust. To quote someone wiser than me:
> My initial prototype here used message-passing type pipes with a custom rolled protocol. If I make my own protocol, it needs to be fuzzed. And I'm probably missing something. Many/most of these concerns for security are probably eliminated if we use a well-known mechanism for this sort of IPC. My thoughts go to a COM server. More complicated to implement than message pipes, but probably brings a lot of security benefits and eliminates the need to fuzz the protocol (probably).